common/state_machine.hpp

```diff
@@ -0,0 +1,231 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_STATE_MACHINE_HPP
+#define MENDER_COMMON_STATE_MACHINE_HPP
+
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <common/events.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace common {
+namespace state_machine {
+
+using namespace std;
+
+namespace events = mender::common::events;
+namespace log = mender::common::log;
+
+template <typename ContextType, typename EventType>
+class StateMachineRunner;
+
+template <typename EventType>
+class EventPoster {
+public:
+	virtual ~EventPoster() {
+	}
+
+	virtual void PostEvent(EventType event) = 0;
+};
```

Comment:
Will there ever be multiple classes implementing this interface? If not, I'd save those lines of code, dynamic dispatch and whatever else for other things.

Reply:
It's pretty unlikely, but this interface is not about that. It's about restricting the view of the objects that the states get as arguments. They should not be allowed to tinker with the state runner. All they should see is an interface where they can post events, and that's it.